### PR TITLE
Implement the `UserAgentInterceptor` for the SDK

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
This PR ports the SDK's user agent `MapRequest` layer into the new interceptor format and wires it up via codegen.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-runtime-api/src/client/interceptors/error.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
 #[derive(Debug)]
 pub struct InterceptorError {
     kind: ErrorKind,
+    context: Option<String>,
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime-api/src/client/interceptors/error.rs:17]: Can/should this be a `Option<Cow<'static, str>>`?

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt [hunk 2]
```diff
@@ -45,9 +49,17 @@ class UserAgentDecorator : ClientCodegenDecorator {
...
     }
 
+    override fun serviceRuntimePluginCustomizations(
+        codegenContext: ClientCodegenContext,
+        baseCustomizations: List<ServiceRuntimePluginCustomization>,
+    ): List<ServiceRuntimePluginCustomization> =
+        baseCustomizations.letIf(codegenContext.settings.codegenConfig.enableNewSmithyRuntime) {
+            it + listOf(AddApiMetadataIntoConfigBag(codegenContext))
+        }
+
```
Comment (rcoh) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt:62]: could we just gate the usage of this on the flag and always emit the customizations? Also, do we want service plugins to be able to delete others? if not, we could refactor this API to remove baseCustomizations

File: aws/rust-runtime/aws-runtime/src/user_agent.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+impl UserAgentInterceptor {
+    /// Creates a new `UserAgentInterceptor`
+    pub fn new() -> Self {
+        UserAgentInterceptor
+    }
+}
+
+fn header_values(ua: &AwsUserAgent) -> Result<(HeaderValue, HeaderValue), InvalidHeaderValue> {
+    // Pay attention to the extremely subtle difference between ua_header and aws_ua_header below...
```
Comment (Velfi) [inline aws/rust-runtime/aws-runtime/src/user_agent.rs:34]: ty for this comment

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt [hunk 4]
```diff
@@ -86,8 +98,34 @@ class UserAgentDecorator : ClientCodegenDecorator {
...
+                        .expect("interceptors set")
+                        .register_client_interceptor(std::sync::Arc::new(#{UserAgentInterceptor}::new()) as _);
+                    """,
+                    "API_METADATA" to ClientRustModule.Meta.toType().resolve("API_METADATA"),
+                    "HttpRequest" to smithyRuntimeApi.resolve("client::orchestrator::HttpRequest"),
+                    "HttpResponse" to smithyRuntimeApi.resolve("client::orchestrator::HttpResponse"),
+                    "Interceptors" to smithyRuntimeApi.resolve("client::interceptors::Interceptors"),
+                    "UserAgentInterceptor" to awsRuntime.resolve("user_agent::UserAgentInterceptor"),
+                )
+            }
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt:122]: Good idea!

File: rust-runtime/aws-smithy-runtime-api/src/client/interceptors/error.rs (hunk 1 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/interceptors/error.rs:17]: I'm not happy with the `InterceptorError` in its current incantation. Going to do some refactoring here.

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt (hunk 2 above)
Comment (jdisanti) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/UserAgentDecorator.kt:62]: > could we just gate the usage of this on the flag and always emit the customizations?

This seems like bikeshedding to me, but I might be misunderstanding.

> Also, do we want service plugins to be able to delete others?

Hard to say right now. Probably not?

**Review by Velfi — APPROVED** [review]
The error macros are quite cool. Good job

**Pull request merged**